File: services/planning/app/metrics.py
"""Middleware Prometheus pour le service planning.

Ce que ce module émet, conforme à la convention RED documentée dans le poly :
    - http_requests_total                  (Counter)
    - http_request_duration_seconds        (Histogram, buckets configurables via env)
    - planning_build_info                  (Gauge constante = 1, métadonnées de release)
    - business_event_total (optionnel)     (Counter, activable par env)

Labels normalisés :
    - method        : GET, POST, PUT, ...
    - route         : pattern de route FastAPI (ex. "/slots/{slot_id}"),
                      JAMAIS l'URL brute, pour ne pas exploser la cardinalité.
    - status_class  : "2xx", "3xx", "4xx", "5xx" — groupé pour la même raison.
"""

from __future__ import annotations

import os
import time
from typing import Callable, List, Tuple

from prometheus_client import (
    CONTENT_TYPE_LATEST,
    CollectorRegistry,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response


VERSION = os.getenv("VERSION", "dev")
COMMIT = os.getenv("COMMIT", "unknown")
BUSINESS_ENABLED = os.getenv("METRICS_BUSINESS_ENABLED", "false").lower() == "true"


def _parse_buckets() -> List[float]:
    """Buckets configurables via env METRICS_BUCKETS = "0.05,0.1,0.2,0.3,0.5,1,2,5".

    Les étudiants alignent ces buckets sur leur SLO de latence en étape 2.
    """
    raw = os.getenv("METRICS_BUCKETS", "0.05,0.1,0.2,0.3,0.5,1,2,5")
    parsed: List[float] = []
    for tok in raw.split(","):
        tok = tok.strip()
        if not tok:
            continue
        try:
            v = float(tok)
            if v > 0:
                parsed.append(v)
        except ValueError:
            continue
    return sorted(parsed) if parsed else [0.05, 0.1, 0.2, 0.3, 0.5, 1, 2, 5]


# On utilise un registry dédié pour éviter les collisions avec le registry global
# du process (utile si plusieurs middlewares étaient instanciés en tests).
REGISTRY = CollectorRegistry(auto_describe=True)

HTTP_REQUESTS_TOTAL = Counter(
    "http_requests_total",
    "Nombre total de requêtes HTTP servies.",
    labelnames=["method", "route", "status_class"],
    registry=REGISTRY,
)

HTTP_REQUEST_DURATION = Histogram(
    "http_request_duration_seconds",
    "Durée des requêtes HTTP en secondes.",
    labelnames=["method", "route", "status_class"],
    buckets=_parse_buckets(),
    registry=REGISTRY,
)

BUILD_INFO = Gauge(
    "planning_build_info",
    "Métadonnées de release du service (version, commit, langage). Vaut toujours 1.",
    labelnames=["version", "commit", "language"],
    registry=REGISTRY,
)
BUILD_INFO.labels(version=VERSION, commit=COMMIT, language="python").set(1)

BUSINESS_EVENT_TOTAL = Counter(
    "business_event_total",
    "Compteur d'événements métier applicatifs (activable via METRICS_BUSINESS_ENABLED).",
    labelnames=["kind"],
    registry=REGISTRY,
)


def _status_class(code: int) -> str:
    if code >= 500:
        return "5xx"
    if code >= 400:
        return "4xx"
    if code >= 300:
        return "3xx"
    if code >= 200:
        return "2xx"
    return "1xx"


def _normalize_route(request: Request) -> str:
    """Récupère le pattern FastAPI matché plutôt que l'URL brute.

    Pour Starlette/FastAPI, après le routing, request.scope["route"].path
    contient le pattern (ex. "/slots/{slot_id}"). Si la route n'est pas
    matchée (404), on retombe sur request.url.path — la cardinalité reste
    alors un signal visible plutôt qu'écrasé.
    """
    route = request.scope.get("route")
    if route is not None and getattr(route, "path", None):
        return route.path
    return request.url.path or "unknown"


class PrometheusMiddleware(BaseHTTPMiddleware):
    """Middleware ASGI qui mesure chaque requête HTTP."""

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        # /metrics est servi par la route normale ; pas de mesure récursive.
        if request.url.path == "/metrics":
            return await call_next(request)

        start = time.perf_counter()
        status_code = 500
        try:
            response = await call_next(request)
            status_code = response.status_code
            return response
        finally:
            elapsed = time.perf_counter() - start
            labels = {
                "method": request.method,
                "route": _normalize_route(request),
                "status_class": _status_class(status_code),
            }
            HTTP_REQUESTS_TOTAL.labels(**labels).inc()
            HTTP_REQUEST_DURATION.labels(**labels).observe(elapsed)


def metrics_endpoint() -> Tuple[bytes, str]:
    """Retourne (payload, content-type) pour le endpoint /metrics."""
    return generate_latest(REGISTRY), CONTENT_TYPE_LATEST


def record_business_event(kind: str) -> None:
    if BUSINESS_ENABLED:
        BUSINESS_EVENT_TOTAL.labels(kind=kind).inc()

File: services/annuaire/src/metrics.js
// Middleware Prometheus pour le service annuaire.
//
// Ce que ce fichier émet, conforme à la convention RED documentée dans le poly :
//   - http_requests_total                  (Counter)
//   - http_request_duration_seconds        (Histogram, buckets configurables via env)
//   - annuaire_build_info                  (Gauge constante = 1, métadonnées de release)
//   - business_event_total (optionnel)     (Counter, activable par env)
//
// Labels normalisés :
//   - method        : GET, POST, PUT, ...
//   - route         : pattern Express (ex. "/students/:id"), JAMAIS l'URL brute,
//                     pour ne pas exploser la cardinalité.
//   - status_class  : "2xx", "3xx", "4xx", "5xx" — groupé pour la même raison.

const client = require('prom-client');

const VERSION = process.env.VERSION || 'dev';
const COMMIT = process.env.COMMIT || 'unknown';
const BUSINESS_ENABLED = (process.env.METRICS_BUSINESS_ENABLED || 'false').toLowerCase() === 'true';

// Buckets configurables via env METRICS_BUCKETS = "0.05,0.1,0.2,0.3,0.5,1,2,5".
// Les étudiants alignent ces buckets sur leur SLO de latence en étape 2.
function parseBuckets() {
  const raw = process.env.METRICS_BUCKETS || '0.05,0.1,0.2,0.3,0.5,1,2,5';
  const parsed = raw
    .split(',')
    .map((s) => parseFloat(s.trim()))
    .filter((n) => Number.isFinite(n) && n > 0)
    .sort((a, b) => a - b);
  return parsed.length > 0 ? parsed : [0.05, 0.1, 0.2, 0.3, 0.5, 1, 2, 5];
}

const registry = new client.Registry();
client.collectDefaultMetrics({ register: registry });

const httpRequestsTotal = new client.Counter({
  name: 'http_requests_total',
  help: 'Nombre total de requêtes HTTP servies.',
  labelNames: ['method', 'route', 'status_class'],
  registers: [registry],
});

const httpRequestDuration = new client.Histogram({
  name: 'http_request_duration_seconds',
  help: 'Durée des requêtes HTTP en secondes.',
  labelNames: ['method', 'route', 'status_class'],
  buckets: parseBuckets(),
  registers: [registry],
});

const buildInfo = new client.Gauge({
  name: 'annuaire_build_info',
  help: "Métadonnées de release du service (version, commit, langage). Vaut toujours 1.",
  labelNames: ['version', 'commit', 'language'],
  registers: [registry],
});
buildInfo.set({ version: VERSION, commit: COMMIT, language: 'nodejs' }, 1);

const businessEventTotal = new client.Counter({
  name: 'business_event_total',
  help: "Compteur d'événements métier applicatifs (activable via METRICS_BUSINESS_ENABLED).",
  labelNames: ['kind'],
  registers: [registry],
});

// Regroupe un code HTTP en classe ("2xx", "3xx", ...).
function statusClass(code) {
  if (code >= 500) return '5xx';
  if (code >= 400) return '4xx';
  if (code >= 300) return '3xx';
  if (code >= 200) return '2xx';
  return '1xx';
}

// Normalise la route observée.
//   - on PRÉFÈRE req.route.path (pattern matché par Express, ex. "/students/:id") ;
//   - à défaut (404 ou routeur monté), on retombe sur req.path,
//     ce qui peut produire de la cardinalité — c'est volontaire pour signaler
//     l'anomalie aux équipes plutôt que de l'écraser.
function normalizeRoute(req) {
  const matched = req.route && req.route.path;
  if (matched) {
    if (req.baseUrl) return req.baseUrl + matched;
    return matched;
  }
  return req.path || 'unknown';
}

function metricsMiddleware(req, res, next) {
  const start = process.hrtime.bigint();
  res.on('finish', () => {
    const route = normalizeRoute(req);
    const labels = {
      method: req.method,
      route,
      status_class: statusClass(res.statusCode),
    };
    httpRequestsTotal.inc(labels);
    const seconds = Number(process.hrtime.bigint() - start) / 1e9;
    httpRequestDuration.observe(labels, seconds);
  });
  next();
}

async function metricsHandler(_req, res) {
  res.setHeader('Content-Type', registry.contentType);
  res.send(await registry.metrics());
}

function recordBusinessEvent(kind) {
  if (BUSINESS_ENABLED) businessEventTotal.inc({ kind });
}

module.exports = { metricsMiddleware, metricsHandler, recordBusinessEvent, registry };
